server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileTreeImpl.java

```diff
@@ -0,0 +1,141 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.qualityprofile;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.sonar.api.rule.RuleKey;
+import org.sonar.api.utils.System2;
+import org.sonar.core.util.stream.MoreCollectors;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.qualityprofile.ActiveRuleDto;
+import org.sonar.db.qualityprofile.OrgActiveRuleDto;
+import org.sonar.db.qualityprofile.QProfileDto;
+import org.sonar.server.exceptions.BadRequestException;
+import org.sonar.server.qualityprofile.index.ActiveRuleIndexer;
+
+import static org.sonar.server.ws.WsUtils.checkRequest;
+
+public class QProfileTreeImpl implements QProfileTree {
+
+  private final DbClient db;
+  private final RuleActivator ruleActivator;
+  private final System2 system2;
+  private final ActiveRuleIndexer activeRuleIndexer;
+
+  public QProfileTreeImpl(DbClient db, RuleActivator ruleActivator, System2 system2, ActiveRuleIndexer activeRuleIndexer) {
+    this.db = db;
+    this.ruleActivator = ruleActivator;
+    this.system2 = system2;
+    this.activeRuleIndexer = activeRuleIndexer;
+  }
+
+  @Override
+  public List<ActiveRuleChange> removeParentAndCommit(DbSession dbSession, QProfileDto profile) {
+    List<ActiveRuleChange> changes = removeParent(dbSession, profile);
+    activeRuleIndexer.commitAndIndex(dbSession, changes);
+    return changes;
+  }
+
+  @Override
+  public List<ActiveRuleChange> setParentAndCommit(DbSession dbSession, QProfileDto profile, QProfileDto parentProfile) {
+    List<ActiveRuleChange> changes = setParent(dbSession, profile, parentProfile);
+    activeRuleIndexer.commitAndIndex(dbSession, changes);
+    return changes;
+  }
+
+  private List<ActiveRuleChange> setParent(DbSession dbSession, QProfileDto profile, QProfileDto parent) {
+    checkRequest(parent.getLanguage().equals(profile.getLanguage()), "Cannot set the profile '%s' as the parent of profile '%s' since their languages differ ('%s' != '%s')",
+      parent.getKee(), profile.getKee(), parent.getLanguage(), profile.getLanguage());
+
+    List<ActiveRuleChange> changes = new ArrayList<>();
+    if (parent.getKee().equals(profile.getParentKee())) {
+      return changes;
+    }
+
+    checkRequest(!isDescendant(dbSession, profile, parent), "Descendant profile '%s' can not be selected as parent of '%s'", parent.getKee(), profile.getKee());
```

Comment:
I didn't find a UT for this check

Reply:
See `set_grandchild_as_parent_fails()`. Note that I didn't refactor these tests :-)